### PR TITLE
Feature/ssm agent

### PR DESCRIPTION
**Pull request by runningman84**
# PR o'clock

## Description

This pull requests implements ssm agent support as described here:
https://github.com/terraform-aws-modules/terraform-aws-eks/issues/224

### Checklist

- [ ] `terraform fmt` and `terraform validate` both work from the root and `examples/eks_test_fixture` directories (look in CI for an example)
- [ ] Tests for the changes have been added and passing (for bug fixes/features)
- [ ] Test results are pasted in this PR (in lieu of CI)
- [x] I've added my change to CHANGELOG.md
- [ ] Any breaking changes are highlighted above

### CODE REVIEW

**Comment by max-rocket-internet** [comment]
I'm convinced all users of this module would be happy to have this enabled by default on every cluster. What do other people think?

File: workers.tf [hunk 1]
```diff
@@ -131,6 +131,11 @@ resource "aws_iam_role_policy_attachment" "workers_AmazonEC2ContainerRegistryRea
   role       = "${aws_iam_role.workers.name}"
 }
 
+resource "aws_iam_role_policy_attachment" "node-AmazonEC2RoleforSSM" {
```
Comment (antonbabenko) [inline workers.tf]: Please rename `node-AmazonEC2RoleforSSM` => `workers_AmazonEC2RoleforSSM`

Comment (runningman84) [reply workers.tf]: done

**Comment by runningman84** [comment]
> I'm not convinced all users of this module would be happy to have this enabled by default on every cluster. What do other people think?

I think this is the preferred solution from AWS instead off ssh keys which should be disabled by default.

**Comment by max-rocket-internet** [comment]
OK but if [AWS themselves are reluctant to put this in by default](https://github.com/awslabs/amazon-eks-ami/issues/127) and it's very easy to add this to your worker nodes currently (that's what `pre_userdata` and `iam_role_id` are for) then it seems a little heavy handed to install and run an extra service for all users of this module.

Maybe other people can chime in.

**Comment by hobbsh** [comment]
I agree with @max-rocket-internet - it should be optional and `pre_userdata` is a fine place to specify - if my vote means anything. It is not really related to this module nor necessary to run an EKS cluster.

**Comment by runningman84** [comment]
Ok I guess this feature is not needed by all users and will therefore close this pull request

**Comment by max-rocket-internet** [comment]
Hey @runningman84 
Here's a much more "k8s" approach to installing ssm-agent: https://github.com/mumoshu/kube-ssm-agent
🙂

**Comment by Jeeppler** [comment]
Actually, I would like to have AWS SSM support. It would be nice to have AWS SSM support in the EKS cluster module. AWS SSM works on node level and not on container level.

The `kube-ssm-agent` raises some problems in my understanding. The biggest issue I see is that kube-ssm-agent mounts several system files into the container to be able to provide the SSM agent functionality and requires privileged access:

```
        securityContext:
          runAsUser: 0
          privileged: true
        volumeMounts:
        - name: dbus
          mountPath: /var/run/dbus
        - name: run-systemd
          mountPath: /run/systemd
        - name: var-run
          mountPath: /var/run
        - name: etc-sudoers
          mountPath: /etc/sudoers.d
``` 

I personally, would much rather like to see the AWS SSM agent installed directly on the node. 

The changes @runningman84 introduced are pretty small. However, I do agree with @max-rocket-internet regarding installing the AWS SSM agent by default. I would like it to be an optional feature. 

Maybe with a module variable `install_aws_ssm_agent = <false|true>`. `false` should be the default.

**Comment by max-rocket-internet** [comment]
I'm not keen in including it in the module, even if it defaults to `false`.

First, I don't think we should be in the business of installing software. Either on the nodes or in the cluster. @brandoconnor and I already discussed it in https://github.com/terraform-aws-modules/terraform-aws-eks/issues/99

I know here we are talking about software on the nodes here, and not in k8s, but IMO it's the same (or even a little worse). If we start this, I'm sure we will get more PRs for other random system management software.

Second, a daemonset is the *proper* way of doing this in k8s.

> The biggest issue I see is that kube-ssm-agent mounts several system files into the container to be able to provide the SSM agent functionality and requires privileged access

This is normal for a daemonset like this. The extra access and privilege is no different to running it on the host. The [Sysdig helm chart](https://github.com/helm/charts/blob/master/stable/sysdig/templates/daemonset.yaml) would be another example exactly like this. Or look at fluentd charts. Also, using a daemonset means that the resources that are used by the container are accounted for by k8s, just like any other container. If the process is run on the host, outside of k8s, then it is neither accounted for or monitored.

If ssm-agent provides such good functionality then I recommend you create a PR to the helm charts repo 🔥

Third, AWS didn't want to include it in their AMI: https://github.com/awslabs/amazon-eks-ami/issues/127

And my forth and final point: It's super easy to add it already, that's what pre_userdata and iam_role_id are for 😃

**Pull request closed without merging**